core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -589,5 +626,59 @@ State committed(Substream winningSubstream) {
 
     // GuardedBy RetriableStream.lock
     boolean closed;
+
+    long bufferNeeded;
+    // setting to true must be GuardedBy RetriableStream.lock
+    boolean bufferLimitExceeded;
+  }
+
+
+  /**
+   * Traces the buffer used by a substream.
+   */
+  class BufferSizeTracer extends ClientStreamTracer {
+    // Each buffer size tracer is dedicated to one specific substream.
+    private final Substream substream;
+
+    BufferSizeTracer(Substream substream) {
+      this.substream = substream;
+    }
+
+    /**
+     * A message is sent to the wire, so its reference would be released if no retry or
+     * hedging were involved. So at this point we have to held the reference of the message longer
+     * for retry, and we need to increment {@code substream.bufferNeeded}.
+     */
+    @Override
+    public void outboundWireSize(long bytes) {
+      if (state.winningSubstream != null || substream.closed) {
+        return;
+      }
+      substream.bufferNeeded += bytes;
+      long bufferNeeded = substream.bufferNeeded;
+      long savedPerRpcBufferUsed = perRpcBufferUsedUpdater.get(RetriableStream.this);
+      while (bufferNeeded > savedPerRpcBufferUsed) {
+        if (perRpcBufferUsedUpdater
+            .compareAndSet(RetriableStream.this, savedPerRpcBufferUsed, bufferNeeded)) {
+          long savedChannelBufferUsed =
+              channelBufferUsed.addAndGet(bufferNeeded - savedPerRpcBufferUsed);
+          if (bufferNeeded > perRpcBufferLimit || savedChannelBufferUsed > channelBufferLimit) {
+            // If the stream is not closed, commit this stream;
+            // otherwise, which is a racy edge case, just do nothing until next hedge/normal retry
+            // triggers this IF statement.
+            synchronized (lock) {
+              if (substream.closed) {
+                return;
+              }
+              // state.substreamClosed(substream) never happen before
+              substream.bufferLimitExceeded = true;
```

Comment:
It seems like more than one stream can have this set to true. That would cause us to call closed() more than once.

Reply:
I don't get it. More than one stream can have this set to true, yes, but only one of them can commit successfully, how would it cause us to call closed() more than once?

Reply:
Every stream is closed. But we must call masterListener.closed() only once. So if more than one of the streams has this boolean set, and then they complete naturally (or are cancelled), then we'll end up calling masterListener.closed() multiple times.

Reply:
Got you. That's true.

Reply:
fixed